File: Engine/MLP.cpp
#include <iostream>
#include <string>
#include <vector>
#include <random>
#include <memory>
#include "Value.cpp"

using namespace std;

class Neuron {
public:
    string nonlin;
    vector<shared_ptr<Value>> W;
    shared_ptr<Value> b;
    
    Neuron(int in_dim, string nonlin = "") : nonlin(nonlin) {
        static random_device rd;  
        static mt19937 gen(rd()); 
        uniform_real_distribution<double> dis(-1.0, 1.0);
        
        for (int i = 0; i < in_dim; ++i) {
            W.push_back(make_shared<Value>(dis(gen)));
        }
        b = make_shared<Value>(dis(gen), "", vector<shared_ptr<Value>>{}, "b");
    }
    
    shared_ptr<Value> operator()(const vector<double>& X) {
        auto logit = b;
        for (size_t i = 0; i < X.size(); ++i) {
            logit = logit + (W[i] * X[i]);
        }
        if (nonlin == "Tanh") return logit->Tanh();
        if (nonlin == "Relu") return logit->Relu();
        return logit;
    }
    
    shared_ptr<Value> operator()(const vector<shared_ptr<Value>>& X) {
        auto logit = b;
        for (size_t i = 0; i < X.size(); ++i) {
            logit = logit + (W[i] * X[i]);
        }
        if (nonlin == "Tanh") return logit->Tanh();
        if (nonlin == "Relu") return logit->Relu();
        return logit;
    }
    
    vector<shared_ptr<Value>> parameters() {
        vector<shared_ptr<Value>> params = W; 
        params.push_back(b); 
        return params;
    }
};

class Layer {
public:
    vector<Neuron> neurons;
    
    Layer(int in_dim, int out_dim, string nonlin = "") {
        for (int i = 0; i < out_dim; ++i) {
            neurons.emplace_back(in_dim, nonlin);
        }
    }
    
    vector<shared_ptr<Value>> operator()(const vector<double>& X) {
        vector<shared_ptr<Value>> outs;
        for(auto& neuron : neurons) {
            outs.push_back(neuron(X)); 
        }
        return outs; 
    }
    
    vector<shared_ptr<Value>> operator()(const vector<shared_ptr<Value>>& X) {
        vector<shared_ptr<Value>> outs;
        for(auto& neuron : neurons) {
            outs.push_back(neuron(X)); 
        }
        
        return outs; 
    }
    
    vector<shared_ptr<Value>> parameters() {
        vector<shared_ptr<Value>> p_t;
        for (auto& neuron : neurons) {
            auto p = neuron.parameters();
            p_t.insert(p_t.end(), p.begin(), p.end());
        }
        return p_t;
    }
};

class MLP {
public:
    vector<Layer> layers;
    
    MLP() {}
    
    void Linear(int in_dim, int out_dim) {
        layers.emplace_back(in_dim, out_dim);
    }
    
    void Tanh() {
        if (!layers.empty()) {
            for (auto& neuron : layers.back().neurons) {
                neuron.nonlin = "Tanh";
            }
        }
    }
    
    void Relu() {
        if (!layers.empty()) {
            for (auto& neuron : layers.back().neurons) {
                neuron.nonlin = "Relu";
            }
        }
    }
    
    vector<shared_ptr<Value>> operator()(const vector<double>& X) {
        if (layers.empty()) return {};
        
        vector<shared_ptr<Value>> current_X = layers[0](X);
        for (size_t i = 1; i < layers.size(); ++i) {
            current_X = layers[i](current_X);
        }
        return current_X;
    }
    
    vector<shared_ptr<Value>> parameters() {
        vector<shared_ptr<Value>> p_t;
        for (auto& layer : layers) {
            auto p = layer.parameters();
            p_t.insert(p_t.end(), p.begin(), p.end());
        }
        return p_t;
    }
    
    void zero_grad() {
        for (auto& p : parameters()) {
            p->grad = 0.0;
        }
    }
    
    void step(double lr = 0.01) {
        for (auto& p : parameters()) {
            p->data -= lr * p->grad;
        }
    }
    
    string repr() {
        if (layers.empty()) return "MLP (Empty)";
        
        string out = "MLP (\n";
        for (size_t i = 0; i < layers.size(); ++i) {
            Layer& layer = layers[i];
            int in_dim = layer.neurons.empty() ? 0 : layer.neurons[0].W.size();
            int out_dim = layer.neurons.size();
            string activation = layer.neurons.empty() || layer.neurons[0].nonlin == "" ? "Linear" : layer.neurons[0].nonlin;
            
            out += "  (Layer " + to_string(i) + "): [in_dim=" + to_string(in_dim) + 
                   ", out_dim=" + to_string(out_dim) + ", activation=" + activation + "]\n";
        }
        out += ") -> Total Parameters: " + to_string(parameters().size());
        return out;
    }
};

File: Python Version/MLP.py
import random
from Value import Value

class Neuron:
    def __init__(self,in_dim,nonlin=None):
        self.nonlin = nonlin
        self.W = [Value(random.uniform(-1,1)) for _ in range(in_dim)]
        self.b = Value(random.uniform(-1,1))
    
    def __call__(self,X):
        logit = sum((w*x for w,x in zip(self.W,X))) + self.b
        if not self.nonlin:
            return logit
        elif self.nonlin == 'Tanh':
            return logit.Tanh()
        elif self.nonlin == 'Relu':
            return logit.Relu()
    
    def parameters(self):
        return self.W + [self.b]

class Layer:
    def __init__(self,in_dim,out_dim,nonlin = None):
        self.neurons = [Neuron(in_dim,nonlin) for _ in range(out_dim)]
    
    def __call__(self,X):
        outs = [neuron(X) for neuron in self.neurons]
        return outs[0] if len(outs) == 1 else outs
    
    def parameters(self):
        p_t = []
        for neuron in self.neurons:
            p = neuron.parameters()
            p_t.extend(p)
        return p_t

class MLP:
    def __init__(self):
        self.layers = []
    
    def Linear(self,in_dim,out_dim):
        self.layers.append(Layer(in_dim,out_dim))
    
    def Tanh(self):
        for neuron in self.layers[-1].neurons:
            neuron.nonlin = 'Tanh'
    
    def Relu(self):
        for neuron in self.layers[-1].neurons:
            neuron.nonlin = 'Relu'
    
    def __call__(self, X):
        for layer in self.layers:
            X = layer(X)
        return X
    
    def parameters(self):
        p_t = []
        for layer in self.layers:
            p = layer.parameters()
            p_t.extend(p)
        return p_t
    
    def zero_grad(self):
        for p in self.parameters():
            p.grad = 0.0
    
    def step(self, lr=0.01):
        for p in self.parameters():
            p.data -= lr * p.grad
    
    def __repr__(self):
        if not self.layers:
            return "MLP (Empty)"
            
        out = ["MLP ("]
        for i, layer in enumerate(self.layers):
            # Extract dimensions from the neurons
            in_dim = len(layer.neurons[0].W)
            out_dim = len(layer.neurons)
            
            # Extract activation from the first neuron (defaults to 'Linear' if None)
            activation = layer.neurons[0].nonlin or "Linear"
            
            out.append(f"  (Layer {i}): [in_dim={in_dim}, out_dim={out_dim}, activation={activation}]")
            
        out.append(f") -> Total Parameters: {len(self.parameters())}")
        return "\n".join(out)